contracts/Governance/VestingToken.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.5;
+
+/// Token vesting contract.
+/// Adapted from Melonport AG <team@melonport.com>
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract VestingToken {
+    using SafeMath for uint256;
+
+    ERC20 public token; // Radicle ERC20 contract
+    address public owner; // deployer; can interrupt vesting
+
+    bool public interrupted; // whether vesting is still possible
+    bool public isVestingStarted; // whether vesting period has begun
+    uint256 public totalVestingAmount; // quantity of vested token in total
+    uint256 public vestingStartTime; // timestamp when vesting is set
+    uint256 public vestingPeriod; // total vesting period in seconds
+    address public beneficiary; // address of the beneficiary
+    uint256 public withdrawn; // quantity of token withdrawn so far
+
+    modifier notInterrupted() {
+        require(!interrupted, "The contract has been interrupted");
+        _;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Only owner can do this");
+        _;
+    }
+
+    modifier onlyBeneficiary() {
+        require(msg.sender == beneficiary, "Only beneficiary can do this");
+        _;
+    }
+
+    modifier vestingStarted() {
+        require(isVestingStarted, "Vesting must be started");
+        _;
+    }
+
+    /// @param _token Address of token being vested
+    constructor(address _token, address _owner) {
+        token = ERC20(_token);
+        owner = _owner;
+    }
+
+    /// @notice Returns the token amount that is currently withdrawable
+    /// @return withdrawable Quantity of withdrawable Radicle asset
+    function withdrawableBalance() public view returns (uint256 withdrawable) {
+        uint256 timePassed = block.timestamp.sub(vestingStartTime);
+
+        if (timePassed < vestingPeriod) {
+            uint256 vested = totalVestingAmount.mul(timePassed) / vestingPeriod;
+            withdrawable = vested.sub(withdrawn);
+        } else {
+            withdrawable = totalVestingAmount.sub(withdrawn);
+        }
+    }
+
+    /// @notice Grant vesting tokens to the beneficiary.
+    ///
+    /// @param ofBeneficiary Address of beneficiary
+    /// @param ofTokenAmount Amount of tokens
+    /// @param ofVestingPeriod Vesting period in seconds from vestingStartTime
+    function grantTokens(
+        address ofBeneficiary,
+        uint256 ofTokenAmount,
+        uint256 ofVestingPeriod
```

Comment:
This function should accept the timestamp of when the vesting period should start.

Reply:
Same question here, are you saying it should be back-dated to the vesting start date? The way we were going to do this afaik is to send the already-vested tokens directly, and only use the contract for the unvested ones. But I guess it will be impossible to calculate it exactly in that case?

Reply:
Yes, back-dating to the exact moment considered to be the release of the token is one part. There's no need for us send the already-vested tokens directly if we design this API right. The other is that a cliff period may end after the release of the token and with this API somebody will need to track it and try to snipe that moment with a message.

Reply:
Done in 73824c9